src/libsyntax/parse/parser/stmt.rs

```diff
@@ -473,4 +492,61 @@ impl<'a> Parser<'a> {
             "this was erroneously allowed and will become a hard error in a future release"
         }).emit();
     }
+
+    fn recover_stmt_local(
+        &mut self,
+        span: Span,
+        attrs: Vec<Attribute>,
+        msg: &str,
+    ) -> PResult<'a, Option<Stmt>> {
+        let local = self.parse_local(attrs.into())?;
+        self.struct_span_err(span, "invalid variable declaration")
+            .span_suggestion_short(
+                span,
+                msg,
+                "let".to_string(),
+                Applicability::MachineApplicable
+            ).emit();
+        return Ok(Some(self.mk_dummy_stmt(span, local)));
+    }
+
+    fn mk_dummy_stmt(&mut self, span: Span, kind: P<Local>) -> Stmt {
+        Stmt {
+            id: DUMMY_NODE_ID,
+            kind: StmtKind::Local(kind),
+            span,
+        }
+    }
+
+    /// Returns `true` if the next token is Brace
+    fn is_next_brace(&self) -> bool {
+        self.look_ahead(1, |t|  t.eq(&token::OpenDelim(token::Brace)))
+    }
+
+    /// Returns `true` if the token is kw::Mut and next one is an Ident
+    fn is_ident_mut(&self) -> bool {
+        self.token.is_keyword(kw::Mut) &&
+            self.look_ahead(1, |t| t.is_ident())
+    }
+
+    /// Returns `true` if the token is sym::var.
+    fn is_var_sym(&self) -> bool {
+        self.token.is_non_raw_ident_where(|ident| ident.name == sym::var) &&
+            self.is_malformed_declr()
+    }
+
+    fn has_equals(&self) -> bool {
+        self.look_ahead(2, |t|  t.eq(&token::Eq))
+    }
+
+    fn is_auto_for_let(&self) -> bool {
+        self.token.is_keyword(kw::Auto) && self.is_malformed_declr()
```

Comment:
I think we can be more permissing here.

So reasoning using https://doc.rust-lang.org/nightly/nightly-rustc/syntax/ast/enum.ExprKind.html `auto` is a contextual keyword. The ways in which it can be a valid statement begin with:

```rust
auto( // function call
auto[ // indexing
auto { // struct literal
//--^ These are all the delimiters so we can just check for `OpenDelim(_)`.
auto. // method call
auto; // drop `auto` on the floor
auto!( // macro call
auto trait // e.g. `auto trait Foo {`
auto binop // e.g. `auto + 1`
auto as // a cast, e.g. `auto as u8`
auto : // a type ascription, e.g. `auto : u8`
```

meanwhile, reasoning via https://doc.rust-lang.org/nightly/nightly-rustc/syntax/ast/enum.PatKind.html and intersecting with the list above, I believe we should return true if the token after `auto` is any identifier (including keyword, because `ref`, `ref mut`, and `mut`) that isn't `trait` because two consecutive identifiers can only be the start of a statement if it is an `auto trait` and it cannot become a valid expression. We could also cover literals, but given that this is an irrefutable matching context, it doesn't seem necessary.

We can apply a similar logic to `var` as well but here we can be unrestricted wrt. identifiers (`var trait` is not a thing).